marlowe/src/Language/Marlowe/Client.hs

```diff
@@ -117,40 +116,69 @@ marlowePlutusContract = do
         let lookups = Constraints.scriptInstanceLookups validatorInstance
         utx <- either (throwing _ConstraintResolutionError) pure (Constraints.mkTx lookups tx)
         submitTxConfirmed utx
-        apply `select` wait `select` auto
+        marlowePlutusContract
     wait = do
         params <- endpoint @"wait" @MarloweParams @MarloweSchema
-        r <- SM.waitForUpdate (mkMarloweClient params)
-        case r of
-            Just (TypedScriptTxOut{tyTxOutData=_currentState}, _txOutRef) -> do
-                apply `select` wait `select` auto
-            Nothing -> pure () -- the contract is closed, no UTxO
+        _ <- SM.waitForUpdate (mkMarloweClient params)
+        marlowePlutusContract
     apply = do
         (params, inputs) <- endpoint @"apply-inputs" @(MarloweParams, [Input]) @MarloweSchema
-        MarloweData{..}  <- applyInputs params inputs
-        case marloweContract of
-            Close -> pure ()
-            _     -> apply `select` wait `select` auto
+        _ <- applyInputs params inputs
+        marlowePlutusContract
+    redeem = mapError (review _MarloweError) $ do
+        (MarloweParams{rolesCurrency}, party, pkh) <-
+            endpoint @"redeem" @(MarloweParams, Party, PubKeyHash) @MarloweSchema
+        case party of
+            PK _ -> do
+                logWarn @String $ "Can't redeem a PubKey party "
+                    <> show party <> ". This is only for Role parties."
+                marlowePlutusContract
+            Role role -> do
+                let address = ScriptAddress (mkRolePayoutValidatorHash rolesCurrency)
+                utxos <- utxoAt address
+                let spendPayoutConstraints tx ref TxOutTx{txOutTxOut} = let
+                        expectedDatumHash = datumHash (Datum $ PlutusTx.toData role)
+                        amount = txOutValue txOutTxOut
+                        in case txOutDatum txOutTxOut of
+                            Just datumHash | datumHash == expectedDatumHash ->
+                                -- we spend the rolePayoutScript address
+                                Constraints.mustSpendScriptOutput ref unitRedeemer
+                                -- and pay to a token owner
+                                    <> Constraints.mustPayToPubKey pkh amount
+                            _ -> tx
+
+                let spendPayouts = Map.foldlWithKey spendPayoutConstraints mempty utxos
+                    constraints = spendPayouts
+                        -- must spend a role token for authorization
+                        <> Constraints.mustSpendValue (Val.singleton rolesCurrency role 1)
+                    -- lookup for payout validator and role payouts
+                    validator = rolePayoutScript rolesCurrency
+                    lookups = Constraints.otherScript validator
```

Comment:
Sorry, what is this `otherScript validator` lookup for? Just trying to get my head around it

Reply:
The constraints resolver needs the actual script to spend payout outputs. It expects the script in ScriptLookups record, which we fill here.